app/controllers/mixins/ems_common_angular.rb

```diff
@@ -759,6 +761,11 @@ def build_credentials(ems, mode)
         ssh_keypair_password = params[:ssh_keypair_password] ? params[:ssh_keypair_password].gsub(/\r\n/, "\n") : ems.authentication_key(:ssh_keypair)
         creds[:ssh_keypair] = {:userid => params[:ssh_keypair_userid], :auth_key => ssh_keypair_password, :save => (mode != :validate)}
       end
+      if ems.kind_of?(ManageIQ::Providers::Openstack::CloudManager) &&
+         ems.supports_authentication?(:ssh_keypair) && params[:ssh_keypair_userid]
+        ssh_keypair_password = params[:ssh_keypair_password] ? params[:ssh_keypair_password].gsub(/\r\n/, "\n") : ems.authentication_key(:ssh_keypair)
```

Comment:
why do we need a *third* copy of the whole `creds[:ssh_keypair] = ...` code?

Especially since all the other places use a openstack or openstack_infra or rhevm condition.

Please unify the three.

Reply:
Updated version of the PR combines the three.